client/client.go

```diff
@@ -144,25 +149,33 @@ func (c *Client) Run(ctx context.Context) error {
 	c.rw.Unlock()
 
 	// Once the client is connected, ctx cancellation is ignored.
-	// We start two goroutines, and terminate the gRPC connection and exit Run when any of them exits:
+	// We start three goroutines, and terminate the gRPC connection and exit Run when any of them exits:
 	// 1. processSupervisorRequests reads requests (status changes and QAN data) from the supervisor and sends them to the channel.
 	//    It exits when the supervisor is stopped.
 	//    When the gRPC connection is terminated on exiting Run, processChannelRequests exits too.
-	// 2. processChannelRequests reads requests from the channel and processes them.
+	// 2. sendActionResults reads action results from action runner and sends them to the channel.
+	//    It exits when the action runner is stopped.
+	//    When the gRPC connection is terminated on exiting Run, sendActionResults exits too.
+	// 3. processChannelRequests reads requests from the channel and processes them.
 	//    It exits when an unexpected message is received from the channel, or when can't be received at all.
-	//    When Run is left, caller stops supervisor, and that allows processSupervisorRequests to exit.
-	// Done() channel is closed when both goroutines exited.
-	oneDone := make(chan struct{}, 2)
+	//    When Run is left, caller stops supervisor FIXME BUT DOES NOT STOP ACTION RUNNER, and that allows processSupervisorRequests to exit.
```

Comment:
There is a data race due to this.

Reply:
> There is a data race due to this.

So the data race is because Run doesn't stop ActionRunner? Am I understand right?
And I can check this with the race detector?

I thought all work fine as `ActionRunner` stops all actions and closes the channel when `applicationContext` is done.

Here I'm close the channel.
https://github.com/percona/pmm-agent/blob/d02139b4dfcb1683ee8952018b1c9497faaebc8e/actions/concurrent_runner.go#L66-L69

And here I pass the application context to all actions.
https://github.com/percona/pmm-agent/blob/d02139b4dfcb1683ee8952018b1c9497faaebc8e/actions/concurrent_runner.go#L96-L104

What am I doing wrong here?

Reply:
https://github.com/percona/pmm-agent/pull/52

Reply:
Fixed in https://github.com/percona/pmm-agent/pull/47/commits/00cf6f21df6e93b3342fb1f05ca1d77936e760ae